zio-http/src/main/scala/zhttp/http/Http.scala

```diff
@@ -395,33 +553,35 @@ sealed trait Http[-R, +E, -A, +B] extends (A => ZIO[R, Option[E], B]) { self =>
    * performance improves quite significantly if no additional heap allocations
    * are required this way.
    */
-  final private[zhttp] def execute(a: A): HExit[R, E, B] =
+  final private[zhttp] def execute(a: A): HExit[R, E, B] = try {
     self match {
 
-      case Http.Empty                 => HExit.empty
-      case Http.Identity              => HExit.succeed(a.asInstanceOf[B])
-      case Succeed(b)                 => HExit.succeed(b)
-      case Fail(e)                    => HExit.fail(e)
-      case Attempt(a)                 =>
+      case Http.Empty                     => HExit.empty
+      case Http.Identity                  => HExit.succeed(a.asInstanceOf[B])
+      case Succeed(b)                     => HExit.succeed(b)
+      case Fail(e)                        => HExit.fail(e)
+      case Die(e)                         => HExit.die(e)
+      case Attempt(a)                     =>
         try { HExit.succeed(a()) }
         catch { case e: Throwable => HExit.fail(e.asInstanceOf[E]) }
-      case FromFunctionHExit(f)       => f(a)
-      case FromHExit(h)               => h
-      case Chain(self, other)         => self.execute(a).flatMap(b => other.execute(b))
-      case Race(self, other)          =>
+      case FromFunctionHExit(f)           => f(a)
+      case FromHExit(h)                   => h
+      case Chain(self, other)             => self.execute(a).flatMap(b => other.execute(b))
+      case Race(self, other)              =>
         (self.execute(a), other.execute(a)) match {
           case (HExit.Effect(self), HExit.Effect(other)) =>
             Http.fromOptionFunction[Any](_ => self.raceFirst(other)).execute(a)
           case (HExit.Effect(_), other)                  => other
           case (self, _)                                 => self
         }
-      case FoldHttp(self, ee, bb, dd) =>
-        self.execute(a).foldExit(ee(_).execute(a), bb(_).execute(a), dd.execute(a))
+      case FoldHttp(self, ee, df, bb, dd) =>
+        self.execute(a).foldExit(ee(_).execute(a), df(_).execute(a), bb(_).execute(a), dd.execute(a))
 
       case RunMiddleware(app, mid) => mid(app).execute(a)
 
       case When(f, other) => if (f(a)) other.execute(a) else HExit.empty
     }
+  } catch { case e: Throwable => HExit.die(e) }
```

Comment:
Can you try out the impact on performance using JMH with this new addition of `try` `catch` blocks. 
I think it might be trivial, however we can run the existing benchmarks, for before and after. 

In either case it would still make a lot of sense to use `try/catch` inside of `FoldHttp` `RunMiddleware` and `When` only. For the rest of the primitives it isn't necessary.

Reply:
Yes, of course.

Before addition of `try-catch` block:

```
Benchmark                                    Mode  Cnt         Score         Error  Units
CookieDecodeBenchmark.benchmarkApp          thrpt   25    449402,800 ±   18266,644  ops/s
HttpCollectEval.benchmarkApp                thrpt   25       141,316 ±       4,107  ops/s
HttpCollectEval.benchmarkBase               thrpt   25       697,519 ±      66,594  ops/s
HttpCombineEval.benchmarkNotFound           thrpt   25     92837,003 ±    3014,925  ops/s
HttpCombineEval.benchmarkOk                 thrpt   25     55426,657 ±     239,629  ops/s
HttpNestedFlatMapEval.benchmarkHttpFlatMap  thrpt   25     47856,657 ±    2023,530  ops/s
HttpRouteTextPerf.benchmarkHttpProgram      thrpt   25     17671,778 ±     297,390  ops/s
HttpRouteTextPerf.benchmarkUIOProgram       thrpt   25     18006,379 ±     790,400  ops/s
ProbeContentTypeBenchmark.benchmarkApp      thrpt   25  30995063,186 ± 1637227,626  ops/s
```

Using one "big" `try-catch` block:

```
Benchmark                                    Mode  Cnt         Score        Error  Units
CookieDecodeBenchmark.benchmarkApp          thrpt   25    467636,151 ±  16704,395  ops/s
HttpCollectEval.benchmarkApp                thrpt   25       144,920 ±      4,909  ops/s
HttpCollectEval.benchmarkBase               thrpt   25       703,706 ±     85,421  ops/s
HttpCombineEval.benchmarkNotFound           thrpt   25     96244,372 ±   1617,870  ops/s
HttpCombineEval.benchmarkOk                 thrpt   25     53580,021 ±   1513,844  ops/s
HttpNestedFlatMapEval.benchmarkHttpFlatMap  thrpt   25     48274,401 ±   1736,878  ops/s
HttpRouteTextPerf.benchmarkHttpProgram      thrpt   25     17922,818 ±    484,919  ops/s
HttpRouteTextPerf.benchmarkUIOProgram       thrpt   25     18615,089 ±    554,197  ops/s
ProbeContentTypeBenchmark.benchmarkApp      thrpt   25  32176802,854 ± 668624,246  ops/s
```

So there is a slight drop in performance. After refactoring to catch defects selectively in [f7a31ec](https://github.com/dream11/zio-http/pull/1083/commits/f7a31ec74e7225d0fb24982f87cb17a48e00b3ea):

```
Benchmark                                    Mode  Cnt         Score         Error  Units
CookieDecodeBenchmark.benchmarkApp          thrpt   25    438614,585 ±   30583,003  ops/s
HttpCollectEval.benchmarkApp                thrpt   25       145,921 ±       5,433  ops/s
HttpCollectEval.benchmarkBase               thrpt   25       701,859 ±      73,418  ops/s
HttpCombineEval.benchmarkNotFound           thrpt   25     89815,039 ±    2659,129  ops/s
HttpCombineEval.benchmarkOk                 thrpt   25     54384,573 ±     939,273  ops/s
HttpNestedFlatMapEval.benchmarkHttpFlatMap  thrpt   25     46828,967 ±    2756,157  ops/s
HttpRouteTextPerf.benchmarkHttpProgram      thrpt   25     17275,047 ±     622,047  ops/s
HttpRouteTextPerf.benchmarkUIOProgram       thrpt   25     18572,489 ±    1355,385  ops/s
ProbeContentTypeBenchmark.benchmarkApp      thrpt   25  30865483,444 ± 1277180,211  ops/s
```

As far as I can see performance back to normal.